### PR TITLE
[llvm][ScheduleDAG] SUnit::biasCriticalPath() does not find the critical path consistently

### PR DESCRIPTION
**Pull request by csstormq**
Patch co-authored by AtariDreams (gfunni234@gmail.com).

Fixes #38037.

[AMDGPU] Update test results to fix build (#92982)

### CODE REVIEW

**Review by arsenm — CHANGES_REQUESTED** [review]

**Comment by csstormq** [comment]
Thanks a lot for your suggestion, @arsenm . I tried to fix these problems. Please review again, thank you.

File: llvm/test/CodeGen/X86/misched-critical-path.ll [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  %v162 = load i8, ptr @uc, align 1
+  %v163 = zext i8 %v162 to i64
+  %v164 = trunc i64 %v163 to i8
+  %v165 = trunc i64 %v161 to i8
+  %pair166 = cmpxchg ptr @ull, i8 %v164, i8 %v165 monotonic monotonic
+  %v166 = extractvalue { i8, i1 } %pair166, 0
+  %v167 = icmp eq i8 %v166, %v164
+  %v168 = zext i1 %v167 to i8
+  %v169 = zext i8 %v168 to i32
+  store i32 %v169, ptr @ui, align 4
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/misched-critical-path.ll:228]: Can you reduce this at all? It's a rather long test for this

Comment (csstormq) [reply llvm/test/CodeGen/X86/misched-critical-path.ll]: I did that as you suggested. Now the test really looks very refreshing. Thank you very much, @arsenm .

File: llvm/test/CodeGen/X86/misched-critical-path.ll [hunk 2]
```diff
@@ -0,0 +1,37 @@
+; RUN: llc < %s -mtriple=x86_64-apple-darwin8 -misched-print-dags -o - 2>&1 > /dev/null | FileCheck %s
+; REQUIRES: asserts
+
+target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128"
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/misched-critical-path.ll:4]: Probably can drop the datalayout

**Review by arsenm — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/X86/misched-critical-path.ll [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ; RUN: llc < %s -mtriple=x86_64-apple-darwin8 -misched-print-dags -o - 2>&1 > /dev/null | FileCheck %s
 ; REQUIRES: asserts
 
-target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128"
+target datalayout = "e-p:64:64:64-i8:8:8-i32:32:32"
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/misched-critical-path.ll:4]: You have the complete triple, so the datalayout at all shouldn't matter

File: llvm/test/CodeGen/X86/misched-critical-path.ll (hunk 2 above)
Comment (csstormq) [reply llvm/test/CodeGen/X86/misched-critical-path.ll]: Thank you very much for your comments.

File: llvm/test/CodeGen/X86/misched-critical-path.ll (hunk 3 above)
Comment (csstormq) [reply llvm/test/CodeGen/X86/misched-critical-path.ll]: I see. Thank you for your patience.

**Review by arsenm — APPROVED** [review]

**Pull request merged**